BYOP and DB2OnCloudSetup has out of date instructions for datasource configuration #2303 (#2449)
IBM/FHIR: docs/src/pages/guides/BringYourOwnPersistence.md

layout: post
title: Bring your own Persistence Layer
description: Learn how to build and test a persistence layer for the IBM FHIR Server
date:   2021-06-01
permalink: /BringYourOwnPersistence/
---

## Overview

> The IBM FHIR Server is a modular Java implementation of the HL7 FHIR specification with a focus on performance and configurability. The IBM FHIR Server ships with a JDBC persistence layer that works with IBM Db2, PostgreSQL and Apache Derby. With this modular design, its possible to add support for other relational databases and/or plug in any other persistence layer.

Reviewer: "its possible" -> "it's possible"

> The IBM FHIR Server is a modular Java implementation of the HL7 FHIR specification with a focus on performance and configurability. The IBM FHIR Server ships with a JDBC persistence layer that works with IBM Db2, PostgreSQL and Apache Derby. With this modular design, its possible to add support for other relational databases and/or plug in any other persistence layer.

Author: ```suggestion
The IBM FHIR Server is a modular Java implementation of the HL7 FHIR specification with a focus on performance and configurability. The IBM FHIR Server ships with a JDBC persistence layer that works with IBM Db2, PostgreSQL and Apache Derby. With this modular design, it's possible to add support for other relational databases and/or plug in any other persistence layer.
```

This document outlines the interfaces that need to be implemented and the behaviors required for the related methods to work with the IBM FHIR Server.

### Interfaces
Persistence layer interfaces are defined in the `fhir-persistence` module.

* [FHIRPersistence](https://github.com/IBM/FHIR/blob/main/fhir-persistence/src/main/java/com/ibm/fhir/persistence/FHIRPersistence.java) defines the contract between the REST layer and the persistence layer.
* [FHIRPersistenceFactory](https://github.com/IBM/FHIR/blob/main/fhir-persistence/src/main/java/com/ibm/fhir/persistence/FHIRPersistenceFactory.java) is the interface for providing instances of FHIRPersistence to the server.

### Configuration
Which persistence layer is used by the server is determined by the `/fhirServer/persistence/factoryClassname` property in `fhir-server-config.json`. There is more detail on the configuration in the [IBM FHIR Server's User Guide](https://ibm.github.io/FHIR/guides/FHIRServerUsersGuide#33-persistence-layer-configuration)

> When the default `com.ibm.fhir.persistence.jdbc.FHIRPersistenceJDBCFactory` is used, the returned FHIRPersistenceJDBCImpl instance will look up a corresponding datasource, by JNDI name, using the combination of the tenant id and datastore id from the request context. Specifically, the connection strategy will use the `fhirServer/persistence/<datastore_id>/jndiName` property in the tenant config, or -- if the property is omitted --it will construct the name via the following pattern:

Reviewer 2: for consistency
```suggestion
When the default `com.ibm.fhir.persistence.jdbc.FHIRPersistenceJDBCFactory` is used, the returned FHIRPersistenceJDBCImpl instance will look up a corresponding datasource, by JNDI name, using the combination of the tenant id and datastore id from the request context. Specifically, the connection strategy will use the `fhirServer/persistence/<datasourceId>/jndiName` property in the tenant config, or -- if the property is omitted --it will construct the name via the following pattern:
```

``` md

> jdbc/fhir_<tenant_id>_<datastore_id>[_ro]

Reviewer 2: for consistency
```suggestion
jdbc/fhir_<tenantId>_<datasourceId>[_ro]
```

If `SingleResourceResult.success` is set to false, the `SingleResourceResult.outcome` should be an OperationOutcome with a list of one or more issues that prevented the success of the operation.

> Note, `generateResourceId` is required to be implemented to generate UUIDs, please the implementation `TimestampPrefixedUUID`.

Reviewer 2: ```suggestion
Note that a `generateResourceId` implementation is required to generate server-assigned resource ids. For performance reasons, recommend the use of `TimestampPrefixedUUID`.
```

On failure, set `MultiResourceResult.success` to false and set `MultiResourceResult.outcome` to an OperationOutcome with one or more issues which indicate the failure.

#### Extended Operations

The IBM FHIR Server supports [extended operations](https://www.hl7.org/fhir/operations.html). The IBM FHIR Server has some operations which use custom persistence interactions: 

| Operation Name | Interfaces to implement |
|----------------|-------------------------|

> | `$healthcheck` | `getHealth()` must be implemented to indicate if the persistence layer is in a healthy state and can access the data store |

Reviewer 2: should either include the parens or not consistently
```suggestion
| `$healthcheck` | `getHealth` must be implemented to indicate if the persistence layer is in a healthy state and can access the data store |
```

| `$erase`  | `erase` must be implemented so that ResourceEraseRecord is returned for each Erase operation.  If not implemented it throws an exception|
| `$export` | `fetchResourcePayloads` must be implemented so that the ResourcePayload is returned as a BinaryStream |

> | `$reindex`| `isReindexSupported` must be true, and if true, it `reindex` must be implemented so that a Resource is updated after ingestion|

Reviewer 2: ```suggestion
| `$reindex`| `isReindexSupported` must be true and, if true, `reindex` must be implemented so that users can apply the latest search configuration to a resource after it is already ingested|
```
